docs/langref/relay_type.rst

```diff
@@ -0,0 +1,222 @@
+tensor shapes in complex ways, such as broadcasting operators or
+:code:`flatten`, allowing Relay to statically reason about the shapes
+in these cases.
+
+A type relation :code:`R` is an n-ary-input, single-output relation over
+types. Namely, :code:`R` specifies a relationship between its argument
+types and outputs `true` if the relationship holds and `false`
+if it fails to hold. Types given to a relation may be incomplete or
+include shape variables, so type inference must assign appropriate
+values to incomplete types and shape variables for necessary relations
+to hold, if such values exist.
+
+For example we can define an identity relation to be:
+
+.. code-block:: prolog
+    Identity(I, I) :- true
+
+Or we can define the relation for :code:`flatten`:
+
+.. code-block:: prolog
+    Flatten(Tensor(sh, bt), O) :-
+      O = Tensor(sh[0], prod(sh[1:]))
+
+If we have a relation like :code:`Broadcast` it becomes possible
+to type operators like :code:`add`:
+
+.. code-block:: python
+    add : fun<t1 : Type, t2 : Type, t3 : Type>(t1, t2) -> t3
+                where Broadcast
+
+The inclusion of :code:`Broadcast` above indicates that the argument
+types and the return type must be tensors where the shape of `t3` is
+the broadcast of the shapes of `t1` and `t2`. The type system will
+accept any argument types and return type so long as they fulfill
+:code:`Broadcast`. Hence, it is usually convenient to type operators
+in Relay by defining a relation specific to that operator that
+encodes all the necessary constraints on the argument types
+and the return type.
+
+Note that the above example relations are written in Prolog-like syntax,
+but currently the relations must be implemented by users in C++
+or Python. More specifically, Relay's type system uses an *ad hoc* solver
+for type relations in which type relations are actually implemented as
+C++ or Python functions that check whether the relation holds and
+imperatively update any shape variables or incomplete types. The
+functions are run as needed (if an input is updated) until one of the following holds:
+
+1. All relations hold (typechecking succeeds).
+2. A relation fails to hold (a type error).
+3. A fixpoint is reached where shape variables or incomplete types
+remain (either a type error or more type annotations may be needed).
+
+Presently all of the relations used in Relay are implemented in C++.
+See the files in `src/relay/op` for examples of relations implemented
+in C++.
+
+See :py:class:`~tvm.relay.ty.TypeRelation` for its definition and documentation.
+
+Incomplete Type
+~~~~~~~~~~~~~~~
+
+A type or portion of a type that is not yet known.
+Only used during type inference: any omitted type annotation is
+replaced by an incomplete type, which will be replaced by another
+type at a later point.
+
+.. note:: Known as a "type variable" in the type checking literature.
+
+See :py:class:`~tvm.relay.ty.IncompleteType` for its definition and documentation.
```

Comment:
Should there be a discussion of the TypeReporter?

Reply:
Is it needed? It's purely an internal mechanism of type checking (granted, so are incomplete types), only used for implementing type relations